### PR TITLE
T-28 - Update Business logic for Account, Manage actions

### PR DESCRIPTION
**Pull request by AntonBessarabov**
Created account service
updated account and manage controllers
added manage index page notifications

### CODE REVIEW

**Comment by AntonBessarabov** [comment]
[T-28 Update Business logic for Account, Manage actions](https://trello.com/c/fGSUcpkb/29-t-28-update-business-logic-for-account-manage-actions)

File: PortfolioKeeper/BusinessLogic/Services/AccountService.cs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+        private readonly IPortfolioService portfolioService;
+        private readonly IMapper mapper;
+
+        public AccountService(
+            AppUserManager userManager,
+            AppSignInManager signInManager,
+            IPortfolioService portfolioService,
+            IMapper mapper)
+        {
+            this.userManager = userManager ?? throw new ArgumentNullException(nameof(userManager));
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AccountService.cs]: throw to new lines

File: PortfolioKeeper/BusinessLogic/Services/AccountService.cs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+            if (checkEmail != null)
+            {
+                result.Add(new ModelStateErrorDto { Key = "Email", Message = "This Email is already taken." });
+
+                return result;
+            }
+
+            var user = mapper.Map<UserDto, AppUser>(userDto);
+
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AccountService.cs]: remove empty line

File: PortfolioKeeper/BusinessLogic/Services/AccountService.cs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+            this.portfolioService = portfolioService ?? throw new ArgumentNullException(nameof(portfolioService));
+            this.mapper = mapper ?? throw new ArgumentNullException(nameof(mapper));
+        }
+
+        public async Task<IEnumerable<ModelStateErrorDto>> LoginAsync(UserDto userDto, bool rememberMe = false)
+        {
+            if (userDto == null)
+                throw new ArgumentNullException(nameof(userDto));
+
+            var result = new List<ModelStateErrorDto>();
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AccountService.cs]: rename to errors

File: PortfolioKeeper/BusinessLogic/Services/AccountService.cs [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+            if (callbackUrl == null)
+                throw new ArgumentNullException(nameof(callbackUrl));
+
+            await userManager.SendEmailAsync(userId,
+                "Confirm your account", "Please confirm your account by clicking <a href=\"" + callbackUrl + "\">here</a>");
+        }
+
+        public async Task SendResetMessageAsync(string userId, string callbackUrl)
+        {
+            if (userId == null)
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AccountService.cs]: nullorwhitespace. same for url

File: PortfolioKeeper/Contracts/Enums/ManageMessageId.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿namespace PortfolioKeeper.Contracts.Enums
+{
+    public enum ManageMessageId
```
Comment (stas-litvinenka) [inline PortfolioKeeper/Contracts/Enums/ManageMessageId.cs]: rename to ManageMessage or  ManageMessageValue. put numbers

File: PortfolioKeeper/Controllers/AccountController.cs [hunk 6]
```diff
@@ -1,37 +1,32 @@
...
-            AppSignInManager signInManager,
-            IPortfolioService portfolioService)
+        public AccountController(
+            IAccountService accountService,
+            IMapper mapper)
         {
-            this.userManager = userManager ?? throw new ArgumentNullException(nameof(userManager));
-            this.signInManager = signInManager ?? throw new ArgumentNullException(nameof(signInManager));
-            this.portfolioService = portfolioService ?? throw new ArgumentNullException(nameof(portfolioService));
+            this.accountService = accountService ?? throw new ArgumentNullException(nameof(accountService));
```
Comment (stas-litvinenka) [inline PortfolioKeeper/Controllers/AccountController.cs]: throw to new lines

File: PortfolioKeeper/Controllers/ManageController.cs [hunk 7]
```diff
@@ -63,72 +64,39 @@ public ActionResult ChangePassword()
...
-
-            if (user != null)
-            {
-                await signInManager.SignInAsync(user, isPersistent: false, rememberBrowser: false);
-            }
+            string userId = User.Identity.GetUserId();
 
-            return RedirectToAction("Index", new { Message = ManageMessageId.SetPasswordSuccess });
-        }
+            if (String.IsNullOrWhiteSpace(userId))
```
Comment (stas-litvinenka) [inline PortfolioKeeper/Controllers/ManageController.cs]: better to use string.IsNullOrWhiteSpace

File: PortfolioKeeper/Controllers/ManageController.cs [hunk 8]
```diff
@@ -63,72 +64,39 @@ public ActionResult ChangePassword()
...
-        public async Task<ActionResult> ChangeProfile()
-        {
-            var user = await userManager.FindByIdAsync(User.Identity.GetUserId());
+            var userDto = await accountService.GetUserByIdAsync(userId);
 
-            if (user == null)
+            if (userDto == null)
             {
-                throw new NotImplementedException();
+                throw new SecurityException(); //todo: ???
```
Comment (stas-litvinenka) [inline PortfolioKeeper/Controllers/ManageController.cs]: put message inside SecurityException. remove todo

File: PortfolioKeeper/BusinessLogic/Services/AccountService.cs [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+            portfolioService.Create(portfolioDto);
+
+            return errors;
+        }
+
+        public async Task SendConfirmMessageAsync(string userId, string callbackUrl)
+        {
+            if (string.IsNullOrWhiteSpace(userId))
+                throw new ArgumentNullException(nameof(userId));
+
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AccountService.cs]: remove line

File: PortfolioKeeper/BusinessLogic/Services/AccountService.cs [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+            return errors;
+        }
+
+        public async Task<IEnumerable<ModelStateErrorDto>> RegisterAsync(UserDto userDto)
+        {
+            if (userDto == null)
+                throw new ArgumentNullException(nameof(userDto));
+
+            var errors = new List<ModelStateErrorDto>();
+
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AccountService.cs]: remove line

File: PortfolioKeeper/BusinessLogic/Services/AccountService.cs [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+        public async Task<IEnumerable<ModelStateErrorDto>> ChangePasswordAsync(UserDto userDto, string newPassword)
+        {
+            if (userDto == null)
+                throw new ArgumentNullException(nameof(userDto));
+
+            if (String.IsNullOrWhiteSpace(newPassword))
+                throw new ArgumentNullException(nameof(newPassword));
+
+            var result = new List<ModelStateErrorDto>();
+
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AccountService.cs]: remove line

File: PortfolioKeeper/BusinessLogic/Services/AccountService.cs [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+            return result;
+        }
+
+        public async Task<IEnumerable<ModelStateErrorDto>> ChangePasswordAsync(UserDto userDto, string newPassword)
+        {
+            if (userDto == null)
+                throw new ArgumentNullException(nameof(userDto));
+
+            if (String.IsNullOrWhiteSpace(newPassword))
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AccountService.cs]: String.IsNullOrWhiteSpace(newPassword) -> string.IsNullOrWhiteSpace(newPassword)

File: PortfolioKeeper/BusinessLogic/Services/AccountService.cs [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+            return result;
+        }
+
+        public async Task<IEnumerable<ModelStateErrorDto>> UpdateProfileAsync(UserDto userDto)
+        {
+            if (userDto == null)
+                throw new ArgumentNullException(nameof(userDto));
+
+            var result = new List<ModelStateErrorDto>();
+
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AccountService.cs]: remove line

**Pull request merged**